### PR TITLE
`with_strided_spans`: only break on token boundaries

### PR DESCRIPTION
**Pull request by danieldk**
## Description

`with_strided_spans` would always break up docs at the stride/window boundaries. However, this could split tokens that consists of multiple pieces. This change resolves that issue by always splitting on the first word boundary after the stride/window.

### Types of change

Enhancement

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: curated_transformers/models/with_strided_spans.py [hunk 1]
```diff
@@ -321,3 +346,50 @@ def _apply_to_layers(input: List[List[Floats2d]]):
         Xlr: List[Ragged] = []
         _apply_to_layer(flat_tuple, Xlr)
         return Xlr
+
+
+def _find_spans_with_token_boundaries(
```
Comment (shadeMe) [inline curated_transformers/models/with_strided_spans.py:351]: Can we add a test for this function that tests all three cases described below?

File: curated_transformers/models/with_strided_spans.py [hunk 2]
```diff
@@ -95,13 +100,11 @@ def with_strided_spans_forward(
     window: int = model.attrs["window"]
     batch_size: int = model.attrs["batch_size"]
 
-    spans, doc_lens = _ragged_to_strided_arrays(Xlr=X, stride=stride, window=window)
-    # Calculate once and use for all layer outputs.
-    doc_len_sums = [int(x.sum()) for x in doc_lens]
+    spans = _ragged_to_strided_arrays(Xlr=X, stride=stride, window=window)
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/models/with_strided_spans.py]: Done.

File: curated_transformers/models/with_strided_spans.py [hunk 3]
```diff
@@ -321,3 +346,50 @@ def _apply_to_layers(input: List[List[Floats2d]]):
...
+        # We find the next boundary by first computing the cumulative
+        # sums of the token lengths. We can then (binary) search the token
+        # that lies on the stride/window boundary. Suppose that we have
+        # a stride of 128, there are three scenarios:
+        #
+        # 1. A particular token has a cumulative length of 128. In this case,
+        #    splitting on the stride wouldn't split up a token.
+        # 2. No token has a cumulative length of 128, but there is a token
+        #    with a cumulative length >128. The binary search will return the
+        #    index of the first token token with length >128, say 130. We
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/models/with_strided_spans.py]: Fixed.

File: curated_transformers/models/with_strided_spans.py [hunk 4]
```diff
@@ -321,3 +346,50 @@ def _apply_to_layers(input: List[List[Floats2d]]):
...
+        #    will then use a stride of 130 to include the full token.
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/models/with_strided_spans.py]: Used your suggestion.

File: curated_transformers/models/with_strided_spans.py (hunk 1 above)
Comment (danieldk) [reply curated_transformers/models/with_strided_spans.py:354]: Added.

**Review by shadeMe — APPROVED** [review]

**Pull request merged**